src/content/en/fundamentals/engage-and-retain/push-notifications/sending-messages.markdown

```diff
@@ -3,93 +3,343 @@ layout: shared/narrow
 
-{% highlight json %}
-{
-  "endpoint": "https://example.com/push-service/send/dbDqU8xX10w:APA91b...",
-  "keys": {
-    "auth": "qLAYRzG9TnUwbprns6H2Ew==",
-    "p256dh": "BILXd-c1-zuEQYXH\_tc3qmLq52cggfqqTr\_ZclwqYl6A7-RX2J0NG3icsw..."
-  }
+The following diagram shows the order of operations.
+
+![Sending a message](images/push-flow.gif)
+
+1. A device downloads your web app containing an already created publicKey, referred to in scripts as the `applicationServerKey`. Your web app installs a service worker.
+1. During the subscription flow the browser posts a request to the messaging server to create a new subscription.
+
+   Note: You don't need to know the URL of the message server. Each browser vender manages it's own message server for its browser.
+
+1. The messaging server returns 201 to indicate that the subscription was successful.
+1. After processing the subscription flow, your app passes a subscription object back to your app server. 
+1. At some later point, your app server sends a message to the messaging server.
+
+## Generating the applicationServerKey
+
+There are several things you need to know about the `applicationServerKey`:
+
+* It's the public key portion of a public/private key pair generated on your application server.
+* The key pair should be usable with an elliptic curve digital signature (ECDSA) over the P-256 curve.
+* Your app should pass key to the server as an array of eight-bit unsigned integers.
+* It's required by a spec called Voluntary Application Server Identification for Web Push (VAPID), which we'll discuss in the [sending messages](sending-messages) section. 
+
+You can find an example for generating this in the [web-push node library](https://github.com/web-push-libs/web-push/). It looks something like this:
+
+{% highlight javascript %}
+function generateVAPIDKeys() {
+  var curve = crypto.createECDH('prime256v1');
+  curve.generateKeys();
+
+  return {
+    publicKey: curve.getPublicKey(),
+    privateKey: curve.getPrivateKey(),
+  };
+}
+{% endhighlight %}
+
+## Anatomy of a subscription object
+
+Earlier we said that a subscription object must be stringified and passed to the 
+server, but we didn't tell you what was in the subscription object. That's 
+because the client doesn't do anything with it. The server does.  
+  
+The subscription object looks like this:  
+
+{% highlight javascript %}
+{  
+  "endpoint": "https://example.com/push-service/send/dbDqU8xX10w:APA91b...",  
+  "keys": {  
+    "auth": "qLAYRzG9TnUwbprns6H2Ew==",  
+    "p256dh": 
+"BILXd-c1-zuEQYXH\\_tc3qmLq52cggfqqTr\\_ZclwqYl6A7-RX2J0NG3icsw..."  
+  }  
 }
 {% endhighlight %}
 
-What's in this?
+What's in this?  
+  
+**endpoint**—Contains two parts: the URL of the messaging service used by the
+subscribing browser followed by unique identifier for the user. This is called a
+subscription ID or  a registration ID. This tells your server how to identify
```

Comment:
Drop the subscription ID or registration ID. Just refer to it as a URL for the web push service and some kind of unique ID.